drivers/mpsl/flash_sync/flash_sync_mpsl.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <errno.h>
+
+#include <mpsl.h>
+#include <mpsl_timeslot.h>
+
+#include "multithreading_lock.h"
+#include "soc_flash_nrf.h"
+
+#define LOG_LEVEL CONFIG_SOC_FLASH_NRF_RADIO_SYNC_MPSL_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(flash_sync_mpsl);
+
+struct mpsl_context {
+	struct k_sem sem;  /* Synchronization semamphore. */
+	mpsl_timeslot_session_id_t session_id; /* Timeslot session ID. */
+	uint32_t duration; /* Requested duration. */
+	/* Argument passed to nrf_flash_sync_exe(). */
+	struct flash_op_desc *op_desc;
+	mpsl_timeslot_request_t timeslot_request;
+	/* Return parameter for the timeslot session. */
+	mpsl_timeslot_signal_return_param_t return_param;
+	int status; /* Return value for nrf_flash_sync_exe(). */
+};
+
+static struct mpsl_context _context;
+
+/* Update _context.return_param with parameters for the next timeslot. */
+static void schedule_next_timeslot(void)
+{
+	_context.return_param.callback_action =
+		MPSL_TIMESLOT_SIGNAL_ACTION_REQUEST;
+	_context.return_param.params.request.p_next =
+		&_context.timeslot_request;
+}
+
+static mpsl_timeslot_signal_return_param_t *timeslot_callback(
+		mpsl_timeslot_session_id_t session_id,
+		uint32_t signal)
+{
+	__ASSERT_NO_MSG(session_id == _context.session_id);
+
+	switch (signal) {
+	case MPSL_TIMESLOT_SIGNAL_START:
+		if (_context.op_desc->handler(_context.op_desc->context)
+				== FLASH_OP_DONE) {
+			_context.status = 0;
+			_context.return_param.callback_action =
+				MPSL_TIMESLOT_SIGNAL_ACTION_END;
+			k_sem_give(&_context.sem);
+		} else {
+			schedule_next_timeslot();
+		}
+
+		break;
+
+	case MPSL_TIMESLOT_SIGNAL_SESSION_IDLE:
+	case MPSL_TIMESLOT_SIGNAL_SESSION_CLOSED:
+	case MPSL_TIMESLOT_SIGNAL_CANCELLED:
```

Comment:
MPSL_TIMESLOT_SIGNAL_CANCELLED should be handled the same way as MPSL_TIMESLOT_SIGNAL_BLOCKED